[serve][docs] Consolidate production guides, add kuberay docs to it (#27747)
ray-project/ray: doc/source/serve/production.md

(serve-in-production)=

# Production Guide

The recommended way to run Ray Serve in production is on Kubernetes using the KubeRay [RayService] custom resource because it automatically handles important production requirements such as health checking, status reporting, failure recovery, and upgrades.

> If you're not running on Kubernetes, you can also run Ray Serve on a single node or on a multi-node cluster (likely managed using the [Ray Cluster Launcher]) using the Serve CLI.

Reviewer: > If you're not running on Kubernetes, you can also run Ray Serve on a single node or on a multi-node cluster (likely managed using the [Ray Cluster Launcher]) using the Serve CLI.

Should [Ray Cluster Launcher] be a link?

This section should help you:

- understand the Serve config file format.
- understand how to deploy Ray Serve using the REST API / CLI directly.

> - understand how to deploy Ray Serve on Kubernetes using the [RayService] custom resource.

Reviewer: > understand how to deploy Ray Serve on Kubernetes using the [RayService] custom resource.

Should [RayService] be a link?

> - understand how to deploy Ray Serve on Kubernetes using the [RayService] custom resource.

Author: we have been using lower case across the pages because the bullets are a continuation of the sentence above. If we change it we'll need to change all of the others. I'm not really opinionated either way but prefer to keep it uniform.

```{contents}
```

(serve-in-production-example)=

## Working example: FruitStand application

> This user guide will use the following Serve application as a working example. It takes in requests containing a list of two values, a fruit name and an amount, and returns the total price for the batch of fruits.

Reviewer: ```suggestion
This guide uses the following Serve application as a working example. The app takes in requests containing a list of two values: a fruit name and an amount. It returns the total price for the batch of fruits.
```

```{literalinclude} ../serve/doc_code/production_fruit_example.py
:language: python
:start-after: __fruit_example_begin__
:end-before: __fruit_example_end__
```

This code is saved in the `fruit.py` file and the final deployment graph that we want to deploy is assigned to the `deployment_graph` variable.
To run and test the application locally, use the `serve run fruit:deployment_graph` command:

```console
# Terminal window 1:

$ ls
fruit.py

$ serve run fruit:deployment_graph
2022-06-21 13:07:01,966  INFO scripts.py:253 -- Deploying from import path: "fruit:deployment_graph".
2022-06-21 13:07:03,774  INFO services.py:1477 -- View the Ray dashboard at http://127.0.0.1:8265
...
2022-06-21 13:07:08,076  SUCC scripts.py:266 -- Deployed successfully.

```

> We can test this graph by opening a new terminal window and making requests with Python's [requests](https://requests.readthedocs.io/en/latest/) library.

Reviewer: Can we update this to discuss the `curl` command instead of the `requests` library? Alternatively, could we show how to make a request with the `requests` library?

```console
# Terminal window 2:

$ curl -H "Content-Type: application/json" -d '{"PEAR": 2}' "http://localhost:8000/"
```

(serve-in-production-config-file)=

## Serve Config Files (`serve build`)

The most minimal config file for a Serve application consists of just an `import_path` (the same path you would pass to `serve run` when testing locally).
This config file can be used with the [serve deploy](serve-in-production-deploying) command CLI or embedded in a [RayService] custom resource in Kubernetes to deploy and update your application in production.
The file is written in YAML and has the following format:

> ```yaml

Reviewer: Can we remove `host` and `port` from this yaml example? They're not included in the 2.0 release.

    ...
```

> The `import_path` is the deployment graph's import path. When you deploy your config file, Serve imports your deployment graph using this path. Similarly, the `runtime_env` is the runtime environment that the deployment graph code is imported from and run in.

Author: done

> The `deployments` section is optional. If it's omitted, Serve launches the deployment graph (and all its deployments). The graph runs with any deployment settings specified in the `@serve.deployment` decorators from the graph's code. If you want to override these decorator settings from the code, you can include a `deployments` section in the file. You can add an entry of deployment settings to the `deployments` list. The only required setting in each list entry is the deployment `name`, which must match one of the deployments from the graph's code.

Author: done, looks way better!

    ...
```

The `import_path` is the deployment graph's import path. When you deploy your config file, Serve will import your deployment graph using this path. Similarly, the `runtime_env` is the runtime environment that the deployment graph code will be imported from and run in.

> `host` and `port` are HTTP options. They determine the host IP address and the port for your Serve application's HTTP proxies. These are optional settings and can be omitted. By default, the `host` will be set to `0.0.0.0` to expose your deployments publicly, and the port will be set to `8000`.

Reviewer: Can we remove this paragraph? `host` and `port` aren't in the 2.0 release.

The `fruit_config.yaml` file contains:

> ```yaml

Reviewer: Can we remove host and port from this yaml example? They're not included in the 2.0 release.

> and `8000` is the default port for Serve's HTTP proxies. You can modify these

Reviewer: Can we remove this paragraph? `host` and `port` aren't included in the 2.0 release.

* It will start a new Serve application if one hasn't already started.
* The Serve application will deploy the deployments from your deployment graph, updated with the configurations from your config file.

It does **not** mean that your Serve application, including your deployments, has already started running successfully. This happens asynchronously as the Ray cluster attempts to update itself to match the settings from your config file. Check out the [next section](serve-in-production-inspecting) to learn more about how to get the current status.

### Adding a runtime environment

If you start Ray and deploy your deployment graph from a directory that doesn't contain the graph code, your deployments will fail to run. This happens because your import path is generally location-dependent. For example, the import path `fruit:deployment_graph` assumes the current directory contains the `fruit.py` module, which contains a `deployment_graph` object.

> To make your config file location-independent, you can push your deployment graph code to [a remote repository and add that repository to your config file's `runtime_env` field](remote-uris). When Serve runs your deployment graph, it will pull the code from the remote repository rather than use a local copy. **This is a best practice** because it lets you deploy your config file from any machine in any directory and share the file with other developers, making it a more standalone artifact.

Reviewer 2: This might suggest that using a local working_dir instead of remote is just bad practice, but using a local `working_dir` actually isn't supported right?  If so, we should call that out here

Author: I updated, PTAL

* It will start a new Serve application if one hasn't already started.
* The Serve application will deploy the deployments from your deployment graph, updated with the configurations from your config file.

It does **not** mean that your Serve application, including your deployments, has already started running successfully. This happens asynchronously as the Ray cluster attempts to update itself to match the settings from your config file. Check out the [next section](serve-in-production-inspecting) to learn more about how to get the current status.

### Adding a runtime environment

> The import path (e.g., `fruit:deployment_graph`) must be importable by Serve at runtime.

Author: reworded to reduce the redundancy

When running locally, this might be in your current working directory.
However, when running on a cluster you also need to make sure the path is importable.

> You can achieve this either by building the code into the cluster's container image or by using a `runtime_env` with a [remote URI](remote-uris) that hosts the code in remote storage.

Reviewer: > You can achieve this either by building the code into the cluster's container image

Is there documentation about a `cluster's container image` that we can link to? New Ray Serve users might not know how to create a Ray cluster that uses container images.

Author: added a link

(serve-in-production-updating)=

> ### Updating the Serve spplication

Reviewer: ```suggestion
### Updating the Serve application
```

* Use `serve config` to check the latest config that your Serve application received. This is its goal state.
* Make lightweight configuration updates (e.g. `num_replicas` or `user_config` changes) by modifying your Serve config file and redeploying it with `serve deploy`.
* Make heavyweight code updates (e.g. `runtime_env` changes) by starting a new Ray cluster, updating your Serve config file, and deploying the file with `serve deploy` to the new cluster. Once the new deployment is finished, switch your traffic to the new cluster.

(serve-in-production-kubernetes)=

## Deploying to Production on Kubernetes

The above sections walk through how to deploy a Serve application manually using the Serve CLI.

> However, the best practice when running on Kubernetes is to use the [RayService] custom resource (CR) that's provided as part of [KubeRay] because it automatically handles important production requirements such as health checking, status reporting, failure recovery, and upgrades.

Reviewer: There should probably be a broader sentence here mentioning that Ray Serve _can_ be run on Kubernetes. Non-Ray users that are using Kubernetes might not realize that's possible. The intro could probably discuss

1. Ray Serve can be run on K8s
2. Why should/shouldn't users run on K8s
3. (Optionally) why use Ray Serve at all/why not just use K8s

Author: done

A [RayService] CR encapsulates a multi-node Ray Cluster and a Serve application that will run on top of it into a single Kubernetes manifest.
Deploying, upgrading, and getting the status of the application can be done using standard `kubectl` commands.
This section will walk through how to deploy, monitor, and upgrade the [`FruitStand` example](serve-in-production-example) from above on Kubernetes.

### Installing the KubeRay operator

> The first step is to install the `KubeRay` operator into your Kubernetes cluster.

Reviewer: Can we add a link somewhere telling users how to start a Kubernetes cluster? Users that are totally new to model serving might not know how to do this.

Author: done

This will create a pod that runs the `KubeRay` controller to manage resources based on the `RayService` CRs you create.

Install the operator using `kubectl apply` and check that the controller pod is running:
```console

> $ kubectl apply -k "github.com/ray-project/kuberay/ray-operator/config/default"

Reviewer: `"github.com/ray-project/kuberay/ray-operator/config/default"` leads to a 404 page when I enter it in the browser. Should it lead to a specific page?

Should we include a commit hash here to permalink to a specific operator? Is there a chance that the operator could get updated and break this example?

Reviewer 3: @shrekris-anyscale this is the pattern used by kubectl to clone the repo which is is fine...

However, this command isn't even on Kuberay docs 
https://ray-project.github.io/kuberay/deploy/installation/

I don't think it will work out of the box for users?

Author: fixed

NAME                            READY   STATUS    RESTARTS   AGE
ray-operator-75dbbf8587-5lrvn   1/1     Running   0          31s
```

For more details, see the [KubeRay installation guide](https://ray-project.github.io/kuberay/deploy/installation/).

### Deploying a Serve application

Once the KubeRay controller is running, you can manage your Ray Serve application by creating and updating a `RayService` CR.
`RayService` custom resources consist of the following:
- a `KubeRay` `RayCluster` config defining the parameters for the cluster that the Serve application will run on.
- a Ray Serve [config](serve-in-production-config-file) defining the Serve application to run on the cluster.

When the `RayService` is created, the `KubeRay` controller will first create a Ray cluster using the provided configuration.
Then, once the cluster is running, it will deploy the Serve application to the cluster using the [REST API](serve-in-production-deploying).
The controller will also create a Kubernetes Service that can be used to route traffic to the Serve application.

Let's see this in action by deploying the same [`FruitStand` example](serve-in-production-example) from above.

> The Serve config for the example is embedded into [this example `RayService` CR](https://github.com/ray-project/kuberay/blob/master/ray-operator/config/samples/ray_v1alpha1_rayservice.yaml).

Reviewer: Should we use a specific commit hash here to prevent accidental breakage if the YAML file is updated?

Reviewer 3: Yeah maybe use this branch? https://github.com/ray-project/kuberay/tree/release-0.3

Author: good catch done

This service can be queried directly from inside the cluster, but to access it from your laptop you'll need to configure a Kubernetes ingress or use port forwarding as below:

```console
$ kubectl port-forward service/rayservice-sample-serve-svc 8000
$ curl -X POST -H 'Content-Type: application/json' localhost:8000 -d '["MANGO", 2]'
6
```

### Getting the status of the application

As the `RayService` is running, the `KubeRay` controller will continually health check and monitor it and write status updates to the CR.
You can view the status of the application using `kubectl describe`.
This will include the status of the cluster, any events that happened such as health check failures or restarts, and the same application-level statuses reported by [`serve status`](serve-in-production-inspecting).

```console
$ kubectl get rayservices
NAME                AGE
rayservice-sample   7s

$ kubectl describe rayservice rayservice-sample
```

### Updating the application

To update the `RayService`, modify the manifest and apply it use `kubectl apply`.
There are two types of updates that can occur:

> - *Application-level updates*: when only the Serve config options are changed, the update will be applied _in-place_ on the same Ray cluster. This enables lightweight updates such as scaling a deployment up or down or modifying autoscaling parameters.

Reviewer: Once #27706 is merged, we should link to that info here to clarify what a "lightweight update" is.

> - *Cluster-level updates*: when the `RayCluster` config options are changed, such as updating the container image for the cluster, it may result in a cluster-level update. In this case, a new cluster will be started, the application will be deployed to it, and once it is running the Kubernetes service will be updated to point at the new cluster. There should not be any downtime for the application.

Reviewer: Should we clarify that this requires enough resources to duplicate the running cluster? Or is that a natural assumption?

    status: UPDATING
...
```

If we query the application, we can see that we now get a different result reflecting the updated price:

```console
$ curl -X POST -H 'Content-Type: application/json' localhost:8000 -d '["MANGO", 2]'
8
```

#### Updating the RayCluster config

The process of updating the RayCluster config is the same as updating the Serve config.
For example, we can update the number of worker nodes to 2 in the manifest:

```console
workerGroupSpecs:
  # the pod replicas in this group typed worker
  - replicas: 2
```

```console
$ kubectl apply -f ray_v1alpha1_rayservice.yaml

$ kubectl describe
...
  pendingServiceStatus:
    appStatus: {}
    dashboardStatus:
      healthLastUpdateTime: "2022-07-18T21:54:53Z"
      lastUpdateTime: "2022-07-18T21:54:54Z"
    rayClusterName: rayservice-sample-raycluster-bshfr
    rayClusterStatus: {}
...
```

In the status, you can see that the `RayService` is preparing a pending cluster.
After the pending cluster is healthy, it will be switched to the active cluster and the previous cluster will be terminated.

## Failure Recovery

Ray Serve is resilient to any component failures within the Ray cluster out of the box.

> You can read more about how process and worker node failures are handled in: {ref}`serve-ft-detail`.

Reviewer: ```suggestion
You can read more about how worker node and process failures are handled in: {ref}`serve-ft-detail`.
```

The other order makes it seem like there's two types of nodes called "process nodes" and "worker nodes."

### Head node failures

By default the Ray head node is a single point of failure: if it crashes, the entire cluster crashes and needs to be restarted.
When running on Kubernetes, the `RayService` controller will health check the cluster and restart it if this occurs, but it will still result in some downtime.
In Ray 2.0, `KubeRay` has added experimental support for [GCS fault tolerance](https://ray-project.github.io/kuberay/guidance/gcs-ha/#ray-gcs-ha-experimental), preventing the Ray cluster from crashing if the head node goes down.
While the head node is recovering, Serve applications can still handle traffic but will not be able to perform any updates or recover from other failures (e.g., actors or worker nodes crashing).
Once the GCS is recovered, the cluster should return to normal behavior.

### Health checking

> By default, each actor making up a Serve deployment is health checked and restarted on failure.

Reviewer: ```suggestion
By default, each Serve deployment replicas is periodically health checked and restarted on failure.
```

:::{note}

> User-defined health checks are experimental and may be subject to change before the interface is stabilized. If you have any feedback or run into any issues or unexpected behaviors, please file an issue on GitHub.

Reviewer: Can we link to the issue page on GitHub?

Author: oh yeah actually this is no longer experimental so i'll remove the note :)

:::

You can customize this behavior to perform an application-level health check or to adjust the frequency/timeout.
To define a custom healthcheck, define a `check_health` method on your deployment class.

> This method should take no arguments and return no result, raising an exception if the replica should be considered unhealthy.

Reviewer: ```suggestion
This method should take no arguments and return no result, and it should raise an exception if the replica should be considered unhealthy.
```

You can also customize how frequently the health check is run and the timeout when a replica will be deemed unhealthy if it hasn't responded in the deployment options.

> ```python
> @serve.deployment(health_check_period_s=10, health_check_timeout_s=30)
> class MyDeployment:
>     def __init__(self, db_addr: str):
>         self._my_db_connection = connect_to_db(db_addr)
>
>     def __call__(self, request):
>         return self._do_something_cool()
>
>     # Will be called by Serve to check the health of the replica.
>     def check_health(self):
>         if not self._my_db_connection.is_connected():
>             # The specific type of exception is not important.
>             raise RuntimeError("uh-oh, DB connection is broken.")
> ```

> [KubeRay]: https://ray-project.github.io/kuberay/

Reviewer: Could we add a header here? These links are floating right now and not very related to the previous section:

<img width="771" alt="Screen Shot 2022-08-10 at 1 39 05 PM" src="https://user-images.githubusercontent.com/92341594/184015672-dac993bf-25d8-4a4f-b26b-1cab441d4810.png">

Author: yeah i'll fix these...

* Use `serve config` to check the latest config that your Serve application received. This is its goal state.
* Make lightweight configuration updates (e.g. `num_replicas` or `user_config` changes) by modifying your Serve config file and redeploying it with `serve deploy`.
* Make heavyweight code updates (e.g. `runtime_env` changes) by starting a new Ray cluster, updating your Serve config file, and deploying the file with `serve deploy` to the new cluster. Once the new deployment is finished, switch your traffic to the new cluster.

(serve-in-production-kubernetes)=

## Deploying to Production on Kubernetes

The above sections walk through how to deploy a Serve application manually to a running Ray cluster using the Serve CLI.
Ray Serve can also be run on Kubernetes, providing the best of both worlds: the user experience and scalable compute of Ray Serve and the production-readiness of Kubernetes.
This also allows you to integrate with existing applications that may be running on Kubernetes.

> The best practice when running on Kubernetes is to use the [RayService] custom resource (CR) that's provided as part of [KubeRay] because it automatically handles important production requirements such as health checking, status reporting, failure recovery, and upgrades.

Reviewer: Why are there names in brackets throughout this description? E.g. `[RayService]` or `[Kuberay]`?

Author: these are links

Author: the link for them is defined at the bottom of the page then you can reuse them

Reviewer: Oh, that's cool, I didn't know that was a feature.

- *Cluster-level updates*: when the `RayCluster` config options are changed, such as updating the container image for the cluster, it may result in a cluster-level update. In this case, a new cluster is started, the application is deployed to it, and once it is running the Kubernetes service is updated to point at the new cluster. There should not be any downtime for the application, but note that this requires the Kubernetes cluster to be large enough to schedule both Ray clusters.

#### Example: Serve config update

In the `FruitStand` example above, let's change the price of a mango in the Serve config to 4:

```console
  - name: MangoStand
    numReplicas: 1
    userConfig: |
      price: 4
```

Now to update the application we apply the modified manifest:

```console
$ kubectl apply -f ray_v1alpha1_rayservice.yaml

$ kubectl describe
...
  serveDeploymentStatuses:
  - healthLastUpdateTime: "2022-07-18T21:51:37Z"
    lastUpdateTime: "2022-07-18T21:51:41Z"
    name: MangoStand

>     status: UPDATING

Reviewer: Does the status actually change to `UPDATING`? After #27000, updating the `user_config` should be a lightweight config update, so the status should continuously be `HEALTHY`. Did any other `MangoStand` field get changed?

Author: even on a user_config change it should briefly get changed to `UPDATING`

Author: though realistically it'll probably be too fast to catch with the CLI. I think it's ok, mostly here for instructive purposes

Reviewer: That sounds good.

...
```

If we query the application, we can see that we now get a different result reflecting the updated price:

```console
$ curl -X POST -H 'Content-Type: application/json' localhost:8000 -d '["MANGO", 2]'
8
```

#### Updating the RayCluster config

The process of updating the RayCluster config is the same as updating the Serve config.

> For example, we can update the number of worker nodes to 2 in the manifest:

Reviewer: > For example, we can update the number of worker nodes to 2 in the manifest:

Can we link `worker nodes` to documentation explaining what a Ray worker node is? I don't think we explain that term anywhere in the Serve docs.

Author: added a link to the clusters key concepts doc.

```console
workerGroupSpecs:

>   # the pod replicas in this group typed worker

Reviewer: > the pod replicas in this group typed worker

Is "group typed worker" a common phrase? I'm not sure what it refers to.

Author: no it's Bruce-language. let me update it :)

  - replicas: 2
```

```console
$ kubectl apply -f ray_v1alpha1_rayservice.yaml

$ kubectl describe
...
  pendingServiceStatus:
    appStatus: {}
    dashboardStatus:
      healthLastUpdateTime: "2022-07-18T21:54:53Z"
      lastUpdateTime: "2022-07-18T21:54:54Z"
    rayClusterName: rayservice-sample-raycluster-bshfr
    rayClusterStatus: {}
...
```

In the status, you can see that the `RayService` is preparing a pending cluster.

> After the pending cluster is healthy, it is switched to the active cluster and the previous cluster is terminated.

Reviewer: ```suggestion
After the pending cluster is healthy, it becomes the active cluster, and the previous cluster is terminated.
```

## Failure Recovery

Ray Serve is resilient to any component failures within the Ray cluster out of the box.
You can read more about how worker node and process failures are handled in: {ref}`serve-ft-detail`.

### Head node failures

By default the Ray head node is a single point of failure: if it crashes, the entire cluster crashes and needs to be restarted.
When running on Kubernetes, the `RayService` controller health checks the cluster and restarts it if this occurs, but this still results in some downtime.
In Ray 2.0, `KubeRay` has added experimental support for [GCS fault tolerance](https://ray-project.github.io/kuberay/guidance/gcs-ha/#ray-gcs-ha-experimental), preventing the Ray cluster from crashing if the head node goes down.
While the head node is recovering, Serve applications can still handle traffic but cannot be updated or recover from other failures (e.g., actors or worker nodes crashing).

> Once the GCS is recovered, the cluster will return to normal behavior.

Reviewer: Can we tell users how to check their head node's state during runtime?

Author: I don't think this is the right place for that, it should be covered in kuberay/cluster management docs

Reviewer: That sounds good

### Health checking

By default, each Serve deployment replica is periodically health checked and restarted on failure.

You can customize this behavior to perform an application-level health check or to adjust the frequency/timeout.
To define a custom healthcheck, define a `check_health` method on your deployment class.

> This method should take no arguments and return no result, and it should raise an exception if the replica should be considered unhealthy.

Reviewer: Can users view this error after failure? If so, can we explain how in this section?

Users might propagate info in the error message to debug why the health check failed, so it would be nice if they could see the message.

Author: I don't think the exception is propagated beyond the logs right now. this would be a good future improvement

Author: i'll make a note

* Use `serve config` to check the latest config that your Serve application received. This is its goal state.
* Make lightweight configuration updates (e.g. `num_replicas` or `user_config` changes) by modifying your Serve config file and redeploying it with `serve deploy`.
* Make heavyweight code updates (e.g. `runtime_env` changes) by starting a new Ray cluster, updating your Serve config file, and deploying the file with `serve deploy` to the new cluster. Once the new deployment is finished, switch your traffic to the new cluster.

(serve-in-production-kubernetes)=

## Deploying to Production on Kubernetes

> The above sections walk through how to deploy a Serve application manually to a running Ray cluster using the Serve CLI.

Reviewer 4: I think it is a bit odd that we start off saying that we recommend using Kubernetes to deploy, but then we jump immediately to walking through the not-so-recommended way. The page is also getting a bit long. Maybe we can:

1. Split out the Kubernetes part into its own page.
2. Make it clear that we will use the CLI/local dev example as a quickstart, but guide the user to the Kubernetes guide for actual production at the beginning/end.

> Ray Serve can also be run on Kubernetes, providing the best of both worlds: the user experience and scalable compute of Ray Serve and the production-readiness of Kubernetes.

Reviewer 4: This makes it sound like Ray Serve is not production-ready :D

Author: hmm yeah I can see that. let me word it a bit more carefully...

This also allows you to integrate with existing applications that may be running on Kubernetes.

> The best practice when running on Kubernetes is to use the [RayService] custom resource (CR) that's provided as part of [KubeRay] because it automatically handles important production requirements such as health checking, status reporting, failure recovery, and upgrades.

Reviewer 4: ```suggestion
The best practice when running on Kubernetes is to use the [RayService] custom resource (CR) that's provided as part of [KubeRay]. The RayService CR automatically handles important production requirements such as health checking, status reporting, failure recovery, and upgrades.
```

A [RayService] CR encapsulates a multi-node Ray Cluster and a Serve application that runs on top of it into a single Kubernetes manifest.
Deploying, upgrading, and getting the status of the application can be done using standard `kubectl` commands.
This section walks through how to deploy, monitor, and upgrade the [`FruitStand` example](serve-in-production-example) from above on Kubernetes.

### Installing the KubeRay operator

This guide assumes that you have a running Kubernetes cluster and have `kubectl` configured to run commands on it.

> See the [Kubernetes documentation](https://kubernetes.io/docs/setup/) if you need help getting started.

Reviewer 4: ```suggestion
See the [Kubernetes documentation](https://kubernetes.io/docs/setup/) or the [KubeRay quickstart guide](kuberay-quickstart) if you need help getting started.
```

The first step is to install the `KubeRay` operator into your Kubernetes cluster.

> This creates a pod that runs the `KubeRay` controller to manage resources based on the `RayService` CRs you create.

Reviewer 4: ```suggestion
This creates a pod that runs the `KubeRay` controller. The `KubeRay` controller manages resources based on the `RayService` CRs you create.
```

Install the operator using `kubectl apply` and check that the controller pod is running:
```console
$ kubectl create -k "github.com/ray-project/kuberay/manifests/cluster-scope-resources?ref=v0.3.0"
$ kubectl apply -k "github.com/ray-project/kuberay/manifests/base?ref=v0.3.0"
$ kubectl get deployments -n ray-system
NAME           READY   UP-TO-DATE   AVAILABLE   AGE
ray-operator   1/1     1            1           40s

$ kubectl get pods -n ray-system
NAME                            READY   STATUS    RESTARTS   AGE
ray-operator-75dbbf8587-5lrvn   1/1     Running   0          31s
```

> For more details, see the [KubeRay installation guide](https://ray-project.github.io/kuberay/deploy/installation/).

Reviewer 4: ```suggestion
For more details, see the [KubeRay quickstart guide](kuberay-quickstart).
```
(probably better if they can stay within docs.ray.io)

Author: oh nice, good idea!

### Deploying a Serve application

> Once the KubeRay controller is running, you can manage your Ray Serve application by creating and updating a `RayService` CR.

Reviewer 4: ```suggestion
Once the KubeRay controller is running, you can manage your Ray Serve application by creating and updating a `RayService` custom resource (CR).
```

`RayService` custom resources consist of the following:

> - a `KubeRay` `RayCluster` config defining the parameters for the cluster that the Serve application runs on.

Reviewer 4: ```suggestion
- a `KubeRay` `RayCluster` config defining the cluster that the Serve application runs on.
```

- a Ray Serve [config](serve-in-production-config-file) defining the Serve application to run on the cluster.

When the `RayService` is created, the `KubeRay` controller first creates a Ray cluster using the provided configuration.
Then, once the cluster is running, it deploys the Serve application to the cluster using the [REST API](serve-in-production-deploying).
The controller also creates a Kubernetes Service that can be used to route traffic to the Serve application.

Let's see this in action by deploying the same [`FruitStand` example](serve-in-production-example) from above.
The Serve config for the example is embedded into [this example `RayService` CR](https://github.com/ray-project/kuberay/blob/release-0.3/ray-operator/config/samples/ray_v1alpha1_rayservice.yaml).

> To follow along, save this CR locally in a file named `ray_v1alpha1_rayservice.yaml`.

Reviewer 4: Provide a bash command to download the file to the right name.

Author: good idea, added

To deploy the example, we simply `kubectl apply` the CR.
This creates the underlying Ray cluster, consisting of a head and worker node pod (see {ref}`_cluster-key-concepts-under-construction` for more details on Ray clusters), as well as the service that can be used to query our application:

```console
$ kubectl apply -f ray_v1alpha1_rayservice.yaml

$ kubectl get rayservices
NAME                AGE
rayservice-sample   7s

$ kubectl get pods
NAME                                                      READY   STATUS    RESTARTS   AGE
ervice-sample-raycluster-qd2vl-worker-small-group-bxpp6   1/1     Running   0          24m
rayservice-sample-raycluster-qd2vl-head-45hj4             1/1     Running   0          24m

$ kubectl get services
NAME                                               TYPE        CLUSTER-IP       EXTERNAL-IP   PORT(S)                                          AGE
kubernetes                                         ClusterIP   10.100.0.1       <none>        443/TCP                                          62d

> # A head node service maintained by the RayService.

Reviewer 4: Not sure that the "maintained by the XXX" parts are important here. Cut?

Author: clarified that the first three are:
`# Services used internally by the KubeRay controller.`

rayservice-sample-head-svc                         ClusterIP   10.100.34.24     <none>        6379/TCP,8265/TCP,10001/TCP,8000/TCP,52365/TCP   24m
# A dashboard agent service maintained by the RayCluster.
rayservice-sample-raycluster-qd2vl-dashboard-svc   ClusterIP   10.100.109.177   <none>        52365/TCP                                        24m
# A head node service maintained by the RayCluster.
rayservice-sample-raycluster-qd2vl-head-svc        ClusterIP   10.100.180.221   <none>        6379/TCP,8265/TCP,10001/TCP,8000/TCP,52365/TCP   24m

> # A serve service maintained by the RayService.

Reviewer 4: ```suggestion
# The Serve service that we will use to send queries to the application.
```

rayservice-sample-serve-svc                        ClusterIP   10.100.39.92     <none>        8000/TCP                                         24m
```

### Querying the application

> Once the `RayService` is running, we can query it over HTTP using the service created by the controller.

Reviewer 4: ```suggestion
Once the `RayService` is running, we can query it over HTTP using the service created by the KubeRay controller.
```

This service can be queried directly from inside the cluster, but to access it from your laptop you'll need to configure a Kubernetes ingress or use port forwarding as below:

```console
$ kubectl port-forward service/rayservice-sample-serve-svc 8000
$ curl -X POST -H 'Content-Type: application/json' localhost:8000 -d '["MANGO", 2]'
6
```

### Getting the status of the application

As the `RayService` is running, the `KubeRay` controller continually monitors it and writes relevant status updates to the CR.
You can view the status of the application using `kubectl describe`.

> This includes the status of the cluster, any events that happened such as health check failures or restarts, and the same application-level statuses reported by [`serve status`](serve-in-production-inspecting).

Reviewer 4: ```suggestion
This includes the status of the cluster, events such as health check failures or restarts, and the application-level statuses reported by [`serve status`](serve-in-production-inspecting).
```

```console
$ kubectl get rayservices
NAME                AGE
rayservice-sample   7s

> $ kubectl describe rayservice rayservice-sample

Reviewer 4: Include example output?

Author: done

```

### Updating the application

To update the `RayService`, modify the manifest and apply it use `kubectl apply`.
There are two types of updates that can occur:
- *Application-level updates*: when only the Serve config options are changed, the update is applied _in-place_ on the same Ray cluster. This enables [lightweight updates](serve-in-production-lightweight-update) such as scaling a deployment up or down or modifying autoscaling parameters.

> - *Cluster-level updates*: when the `RayCluster` config options are changed, such as updating the container image for the cluster, it may result in a cluster-level update. In this case, a new cluster is started, the application is deployed to it, and once it is running the Kubernetes service is updated to point at the new cluster. There should not be any downtime for the application, but note that this requires the Kubernetes cluster to be large enough to schedule both Ray clusters.

Reviewer 4: What happens to the old cluster? How do we delete it?

Reviewer 4: Oh, I see it later on, I'll copy it over here.

> - *Cluster-level updates*: when the `RayCluster` config options are changed, such as updating the container image for the cluster, it may result in a cluster-level update. In this case, a new cluster is started, the application is deployed to it, and once it is running the Kubernetes service is updated to point at the new cluster. There should not be any downtime for the application, but note that this requires the Kubernetes cluster to be large enough to schedule both Ray clusters.

Reviewer 4: ```suggestion
- *Cluster-level updates*: when the `RayCluster` config options are changed, such as updating the container image for the cluster, it may result in a cluster-level update. In this case, a new cluster is started, and the application is deployed to it. Once the new cluster is ready, the Kubernetes service is updated to point to the new cluster and the previous cluster is terminated. There should not be any downtime for the application, but note that this requires the Kubernetes cluster to be large enough to schedule both Ray clusters.
```

8
```

#### Updating the RayCluster config

The process of updating the RayCluster config is the same as updating the Serve config.
For example, we can update the number of worker nodes to 2 in the manifest:

```console
workerGroupSpecs:
  # the number of pods in the worker group.
  - replicas: 2
```

```console
$ kubectl apply -f ray_v1alpha1_rayservice.yaml

$ kubectl describe
...
  pendingServiceStatus:
    appStatus: {}
    dashboardStatus:
      healthLastUpdateTime: "2022-07-18T21:54:53Z"
      lastUpdateTime: "2022-07-18T21:54:54Z"
    rayClusterName: rayservice-sample-raycluster-bshfr
    rayClusterStatus: {}
...
```

In the status, you can see that the `RayService` is preparing a pending cluster.
After the pending cluster is healthy, it becomes the active cluster and the previous cluster is terminated.

## Failure Recovery

Ray Serve is resilient to any component failures within the Ray cluster out of the box.
You can read more about how worker node and process failures are handled in: {ref}`serve-ft-detail`.

### Head node failures

By default the Ray head node is a single point of failure: if it crashes, the entire cluster crashes and needs to be restarted.

> When running on Kubernetes, the `RayService` controller health checks the cluster and restarts it if this occurs, but this still results in some downtime.

Reviewer 4: ```suggestion
When running on Kubernetes, the `RayService` controller health-checks the cluster and restarts it if this occurs, but this still results in some downtime.
```

In Ray 2.0, `KubeRay` has added experimental support for [GCS fault tolerance](https://ray-project.github.io/kuberay/guidance/gcs-ha/#ray-gcs-ha-experimental), preventing the Ray cluster from crashing if the head node goes down.
While the head node is recovering, Serve applications can still handle traffic but cannot be updated or recover from other failures (e.g., actors or worker nodes crashing).
Once the GCS is recovered, the cluster will return to normal behavior.

### Health checking

> By default, each Serve deployment replica is periodically health checked and restarted on failure.

Reviewer 4: ```suggestion
By default, each Serve deployment replica is periodically health-checked and restarted on failure.
```
